### PR TITLE
feat(eval): feature extractor comparison for issue #193

### LINKED ISSUE #193: Evaluate current feature extraction in the robot environment and clarify if and how the feature extraction could be improved.
## Context
This issue originated from #13 and currently combines three different work types:
1. Feature-extractor audit/evaluation
2. New extractor implementation
3. Large-scale training + statistical reporting (possibly with SLURM)

That combination is too broad for one safe PR.

## Problem
Current feature extraction is custom and may be slower / less interoperable than alternative extractor designs. We need a decision on execution strategy before implementation.

## In scope (for this issue right now)
- Define the decision path for how to proceed (evaluation-first vs implementation-first).
- Produce implementation-ready acceptance criteria for the chosen path.

## Out of scope (until decision)
- Full extractor redesign rollout in one PR.
- Full training campaign execution on this machine.
- Unbounded architecture migration touching all observation pipelines.

### Option A: Evaluation-first (Recommended)
Do a benchmark-grade extractor audit first (no behavior-breaking changes), then split implementation into scoped follow-up issues.

Pros:
- Lowest regression risk
- Fastest route to reliable decision data
- Keeps legacy extractor fully intact

Cons:
- No immediate performance win in runtime code

Risk profile: lower speed, highest reliability

### Option B: Implement one new extractor immediately
Add one candidate extractor now and wire it into training scripts.

Pros:
- Immediate feature progress
- Early practical feedback

Cons:
- Higher risk of rework without prior evaluation protocol
- Harder to interpret results if baseline contract is not fixed first

Risk profile: medium speed, medium reliability

### Option C: Full extractor program in one issue
Audit + implement multiple extractors + multi-run statistical analysis in one issue.

Pros:
- Single umbrella execution

Cons:
- Too broad for reviewability and regression control
- High integration risk and timeline uncertainty

Risk profile: highest speed pressure, lowest reliability

## Recommended path
Choose **Option A**.

If approved, this issue should be converted into a narrow evaluation contract and implementation moved to explicit follow-up issues (one issue per extractor family / rollout step).

## Decision needed from maintainer
Please choose one:
- [ ] Option A (evaluation-first, recommended)
- [ ] Option B (single concrete extractor now)
- [x] Option C (full program in one issue)

## Validation contract (once decision is made)
At minimum:
- `BASE_REF=origin/main scripts/dev/pr_ready_check.sh`
- Targeted extractor tests and reproducible training/eval commands added to issue body

### PR DESCRIPTION
**Pull request by ll7**
## Summary

Closes #193. This branch evaluates feature-extractor alternatives for Robot SF PPO, expands the initial 32 K comparison into an Optuna/12M fixed-candidate study, and records the final decision: `dyn_large_med` is the preferred architecture candidate, but it is **not promotion-ready** because hold-out collision rates remain too high.

## Linked Issues

- Closes #193
- Follow-up: #850 investigates collision failures before any default promotion.

## What Changed

- `scripts/tools/benchmark_feature_extractors.py` — reproducible feature-extractor throughput benchmark.
- `configs/scenarios/multi_extractor_eval_193.yaml` — reproducible initial multi-extractor comparison config.
- `robot_sf/training/hardware_probe.py` — fixes CUDA metadata extraction by using `torch.version.cuda` safely.
- `robot_sf/feature_extractors/lstm_extractor.py` plus config/registry wiring — adds LSTM extractor coverage for the study.
- `scripts/training/optuna_feature_extractor.py` and fixed-candidate runner/SLURM helpers — support distributed feature-extractor evaluation.
- `configs/training/ppo/feature_extractor_sweep_base.yaml` and `configs/training/ppo/feature_extractor_candidates_12m_issue193.yaml` — config-first sweep and 12M candidate matrix.
- `docs/context/issue_193_feature_extractor_evaluation.md` and `docs/context/issue_193_feature_extractor_optuna_study.md` — full evidence trail, caveats, Slurm provenance, final recommendation, and follow-up boundary.

## Final Result

The initial 32 K result favored `mlp_small`, but that recommendation is superseded by the later 4M Optuna and 12M fixed-candidate evidence.

12M fixed-candidate training summary:

| Family | Seeds | Mean return | Range | Mean FPS |
|---|---:|---:|---:|---:|
| `dyn_large_med` | 3 | `52.062` | `49.698-54.117` | `768.4` |
| `lc_small_med` | 4 | `45.151` | `32.234-50.178` | `830.6` |
| `dyn_default` | 3 | `25.619` | `21.413-32.526` | `770.8` |

Hold-out policy-analysis result:

| Candidate | Episodes | Success | Collision | Ped collision | Obstacle collision | Path efficiency |
|---|---:|---:|---:|---:|---:|---:|
| `dyn_large_med_s231` | 66 | `0.727` | `0.273` | `0.091` | `0.182` | `0.913` |
| `dyn_large_med_s1337` | 66 | `0.727` | `0.273` | `0.091` | `0.182` | `0.916` |
| `dyn_large_med_s123` | 66 | `0.667` | `0.333` | `0.076` | `0.258` | `0.935` |
| `dyn_default_s1337` | 66 | `0.424` | `0.576` | `0.364` | `0.212` | `0.985` |
| `lc_small_med_s231` | 66 | `0.409` | `0.500` | `0.197` | `0.303` | `0.976` |

## Recommendation

- Treat `dyn_large_med` as the best architecture candidate from issue #193.
- Do **not** promote it as a default yet.
- Do **not** run another broad feature-extractor sweep now.
- Continue in #850 with collision-focused policy work, especially obstacle/wall collisions in bottleneck and merging scenarios.

## Validation / Proof

Commands and artifacts are recorded in `docs/context/issue_193_feature_extractor_optuna_study.md`.

Recent validation evidence:

- `sacct -j 12106 --format=JobID,JobName%18,State,ExitCode,Partition,Elapsed,Start,End -P` — all five hold-out policy-analysis jobs completed with `0:0`.
- `output/benchmarks/issue193_policy_analysis_*/summary.json` — hold-out summaries used for the final table above.
- `uv run ruff check scripts/tools/benchmark_feature_extractors.py docs/context/issue_193_feature_extractor_evaluation.md docs/context/issue_193_feature_extractor_optuna_study.md` — passed after review-comment fixes.
- `uv run python scripts/tools/benchmark_feature_extractors.py --device cpu --batch 2 --warmup 0 --reps 1 --ray-shape 1,272 --drive-shape 1,7 --out output/tmp/bench_fe_smoke_issue193.json` — passed.
- CLI validation checks for `--batch 0`, `--reps 0`, and `--ray-shape 1,0` now fail fast with argparse errors.

Full PR readiness should be run on Slurm/compute resources before merge because this checkout is on the `auxme-imech192` login node.

## Risks / Rollout

- This PR should not change PPO defaults.
- The final recommendation is conservative: architecture selection is complete, but safety/collision promotion is deferred.
- `lightweight_cnn` CUDA nondeterminism remains documented and should not be compared as bitwise reproducible evidence.

## Reviewer Notes

- The key review question is no longer “should `mlp_small` become default?” The answer is no, that was superseded.
- The key result is: `dyn_large_med` wins the architecture comparison, but hold-out collision rates reject promotion.
- Follow-up #850 owns mitigation work for the collision failures.

## Release Notes

* **New Features**
  * Added LSTM feature extractor for sequence-based processing
  * Introduced feature extractor benchmarking tool
  * Added Optuna-based hyperparameter sweep for feature extractor tuning
  * Added fixed-candidate evaluation framework for feature extractors
  * Improved SLURM job submission via array-based scheduling

* **Bug Fixes**
  * Fixed hardware metadata collection for torch.version

* **Documentation**
  * Added feature extractor evaluation guides and results
  * Added SLURM workflow documentation

* **Tests**
  * Added LSTM extractor and hardware profile tests

### CODE REVIEW

**Comment by ll7** [comment]
Currently queued the feature extraction sweep in slurm.

**Comment by ll7** [comment]
Addressed the latest review threads in `4f179aec`:

- Added the missing markdown language fence and clarified that the earlier 32K recommendation is superseded by the later Optuna / 12M study.
- Made the feature-extractor benchmark use `dynamics_original` as the explicit required baseline instead of silently falling back to the first successful extractor.
- Added parse-time validation for non-positive benchmark sizes / shapes and removed the non-ASCII multiplication glyph from benchmark docs/output.
- Fixed the LSTM extractor lint findings and added focused coverage for LSTM extractor variants and hardware-probe CUDA branches.
- Made Slurm-hosted tests robust to Slurm CPU env vars and to missing/old Node runtimes for the ESM validation smoke.

Validation:

- `uv run python scripts/tools/benchmark_feature_extractors.py --help`
- `uv run python scripts/tools/benchmark_feature_extractors.py --reps 0 --device cpu` (expected argparse failure)
- `uv run python scripts/tools/benchmark_feature_extractors.py --batch 0 --device cpu` (expected argparse failure)
- `uv run python scripts/tools/benchmark_feature_extractors.py --ray-shape 1,0 --device cpu` (expected argparse failure)
- `uv run python scripts/tools/benchmark_feature_extractors.py --device cpu --batch 2 --warmup 0 --reps 1 --ray-shape 1,272 --drive-shape 1,7 --out output/tmp/bench_fe_smoke_issue193.json`
- `uv run pytest -q tests/test_feature_extractors.py::TestFeatureExtractors tests/test_feature_extractors.py::TestParameterCounting tests/unit/test_hardware_profile_capture.py`
- Slurm PR readiness: job `12131`, `PYTEST_NUM_WORKERS=4 BASE_REF=origin/main scripts/dev/pr_ready_check.sh` on `a30`, completed successfully with `2915 passed, 19 skipped, 3 warnings in 635.99s`; changed-file coverage has warnings only (`config.py` 96.4%, `hardware_probe.py` 94.1%) and no failures.

**Comment by ll7** [comment]
Synced the PR branch with current `origin/main` in `d04d3a35` and resolved the only merge conflict in `configs/validation/minimal.yaml` by keeping main's `../..` path spelling.

Post-sync validation:

- Slurm PR readiness: job `12133`, `PYTEST_NUM_WORKERS=4 BASE_REF=origin/main scripts/dev/pr_ready_check.sh` on `a30`, completed successfully with `2943 passed, 18 skipped, 3 warnings in 614.37s`.
- Changed-file coverage after sync: `lstm_extractor.py` 100.0%, `hardware_probe.py` 94.1% warning only, `config.py` 96.4% warning only.
- Docstring TODO gate: clean.

**Comment by ll7** [comment]
Addressed the two follow-up review comments in `b44b5268`:

- Switched LSTM ray sequence conversion from `view` to `reshape` and added a regression test with a non-contiguous ray tensor.
- Tightened benchmark shape parsing so `--ray-shape` / `--drive-shape` must be exactly two comma-separated positive integers.

Validation after these fixes:

- `uv run pytest -q tests/test_feature_extractors.py::TestFeatureExtractors::test_lstm_extractor_forward_accepts_non_contiguous_rays tests/test_feature_extractors.py::TestFeatureExtractors::test_lstm_extractor_forward tests/test_feature_extractors.py::TestFeatureExtractors::test_all_extractors_same_interface`
- `uv run ruff check robot_sf/feature_extractors/lstm_extractor.py scripts/tools/benchmark_feature_extractors.py tests/test_feature_extractors.py`
- `uv run ruff format --check robot_sf/feature_extractors/lstm_extractor.py scripts/tools/benchmark_feature_extractors.py tests/test_feature_extractors.py`
- `uv run python scripts/tools/benchmark_feature_extractors.py --ray-shape 272 --device cpu` (expected argparse failure)
- Final Slurm PR readiness: job `12134`, `PYTEST_NUM_WORKERS=4 BASE_REF=origin/main scripts/dev/pr_ready_check.sh`, completed successfully with `2943 passed, 19 skipped, 3 warnings in 617.44s`; changed-file coverage warnings only.
